File: addon/Combos.lua
local _, WS = ...

-- Discovery runtime: activates weapon-weapon synergies (Data\Combos.lua) the
-- moment both weapons are carried, announces them, and records first-time
-- finds in the permanent codex.
WS.ComboSystem = {}
local ComboSystem = WS.ComboSystem

-- Called after any weapon is added to the survivor.
function ComboSystem:Check(player)
    for i = 1, #WS.ComboOrder do
        local id = WS.ComboOrder[i]
        if not player.combosActive[id] then
            local combo = WS.Combos[id]
            local w1 = WS.Player:GetWeapon(player, combo.weapons[1])
            local w2 = WS.Player:GetWeapon(player, combo.weapons[2])
            if w1 and w2 then
                self:Activate(player, id, combo, w1, w2)
            end
        end
    end
end

function ComboSystem:Activate(player, id, combo, w1, w2)
    player.combosActive[id] = true
    -- apply reads its tunable magnitudes from the combo table (3rd arg).
    combo.apply(w1, w2, combo)

    local firstEver = not WoWSurvivorsDB.statistics.discoveries[id]
    WoWSurvivorsDB.statistics.discoveries[id] = true

    WS.Audio:Play("evolve")
    WS.UI:Shake(4, 0.3)
    -- Descriptions may carry {field} tokens (like blessings and class perks), so
    -- a discovery's real magnitudes show instead of the raw placeholder.
    WS.Game:Announce("Discovery: " .. combo.name .. "!",
        WS.FormatTokens(combo.description, combo), 4.0)
    if firstEver then
        WS.UI:Toast(w1.data.icon, "New Discovery!", combo.name)
    end
    WS.Projectile:SpawnFlash(player.x, player.y, 70, WS.Constants.COLORS.arcane)
end

File: src/game/combos.js
/* Discoveries: the moment a survivor carries both weapons of a pair, the combo
 * activates for the run, toasts, and is written into the permanent codex. */
'use strict';
(function (WS) {

  const ComboSystem = {};

  ComboSystem.check = function (p) {
    for (const id of WS.ComboOrder) {
      if (p.combosActive[id]) continue;
      const combo = WS.Combos[id];
      const w1 = WS.Player.getWeapon(p, combo.weapons[0]);
      const w2 = WS.Player.getWeapon(p, combo.weapons[1]);
      if (!w1 || !w2) continue;

      p.combosActive[id] = true;
      combo.apply(w1, w2, combo);
      /* EACH ONE TAKES THE OTHER'S COLOUR.
       *
       * A discovery changed what a weapon did and never changed what it
       * looked like. Frostfire made Cinderfall chill whatever survived it and
       * left the bolt looking exactly like a Cinderfall; the player was told
       * about the pairing once, by a toast, and then never saw it again on
       * the field. Nine discoveries, none of them visible.
       *
       * Derived from the pair rather than written into each entry, so it is
       * one rule for all nine and for every one the game ever gains: a
       * combined weapon carries a core of its partner's colour inside its own
       * shape. The shape stays the weapon's, because that is how the player
       * knows what is firing; the colour inside it says what it is now part
       * of. A discovery that sets its own `blend` keeps it. */
      if (!w1.mods.blend) w1.mods.blend = WS.Weapon.colour(w2);
      if (!w2.mods.blend) w2.mods.blend = WS.Weapon.colour(w1);
      const fresh = WS.Save.discoverCombo(id);
      WS.Game.toast(
        (fresh ? 'Discovery: ' : 'Synergy: ') + combo.name,
        WS.template(combo.description, combo),
        { kind: 'discovery', art: fresh ? 'book' : 'arcane' });
      WS.Audio.play('evolve');
    }
  };

  WS.ComboSystem = ComboSystem;

})(window.WS);
